source/fundamentals/gridfs.txt

```diff
@@ -0,0 +1,387 @@
+how to split files into chunks when storing them and reassemble them when
+retrieving them. The driver implementation of GridFS is an abstraction that
+manages the operations and organization of the file storage.
+
+You should use GridFS if the size of your files exceed the BSON document
+size limit of 16MB. For more detailed information on whether GridFS is
+suitable for your use case, see the :manual:`GridFS server manual page </core/gridfs>`.
+
+See the following sections that describe GridFS operations and how to
+perform them:
+
+- :ref:`Create a GridFS bucket <gridfs-create-bucket>`
+- :ref:`Store Files <gridfs-store-files>`
+- :ref:`Retrieve File Information <gridfs-retrieve-file-info>`
+- :ref:`Download Files <gridfs-download-files>`
+- :ref:`Rename Files <gridfs-rename-files>`
+- :ref:`Delete Files <gridfs-delete-files>`
+- :ref:`Delete a GridFS bucket <gridfs-delete-bucket>`
+
+How GridFS Works
+----------------
+
+GridFS organizes files in a **bucket**, a group of MongoDB collections
+that contain the chunks of files and information describing them. The
+bucket contains the following collections, named using the convention
+defined in the GridFS specification:
+
+- The ``chunks`` collection stores the binary file chunks.
+- The ``files`` collection stores the file metadata.
+
+When you create a new GridFS bucket, the driver creates the collections
+referenced above, prefixed with the default bucket name ``fs``, unless
+you specify a different name. The driver also creates an index on each
+collection to ensure efficient retrieval of the files and related
+metadata. For more information on GridFS indexes, see the server manual
+page on :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`.
+
+When storing files with GridFS, the driver splits the files into smaller
+chunks, each represented by a separate document in the ``chunks`` collection.
+It also creates a document in the ``files`` collection that contains
+a file id as an ``ObjectId``, file name, and other file metadata. You can
+upload the file from memory or from a stream. See the following diagram to
+see how GridFS splits the files when uploaded to a bucket.
+
+.. figure:: /includes/figures/GridFS-upload.png
+   :alt: A diagram that shows how GridFS uploads a file to a bucket
+
+When retrieving files, GridFS fetches the metadata from the ``files``
+collection in the specified bucket and uses the information to reconstruct
+the file from documents in the ``chunks`` collection. You can read the file
+into memory or output it to a stream.
+
+.. _gridfs-create-bucket:
+
+Create a GridFS Bucket
+----------------------
+
+To use GridFS, first create a bucket or reference an existing one on
+a MongoDB database. Call the ``GridFSBuckets.create()`` helper method with
+a database instance as the parameter to instantiate a ``GridFSBucket``.
+You can use the ``GridFSBucket`` instance to call read and write operations on
+the files in your bucket.
```

Comment:
Perhaps I just didn't read this section closely enough, but I was a bit confused at how ``GridFSBuckets.create(database)`` actually works. I guess it's sort of like databases in MongoDB, which you normally just lazily create when you insert documents into them? Or does it actually explicitly instantiate a GridFS bucket, but grab a reference to the existing bucket if it already exists? Maybe we should call out that behavior more explicitly.

Reply:
Good idea.
I wasn't certain whether it was an important detail since we don't call this out for other lazily-created entities like Databases and Collections. While it might be beyond the level of behind-the-scenes type information we usually include, I'll add it since that info probably isn't available anywhere else.